Feature/improve docs (#5)
AitoDotAI/aito-python-tools: docs/source/quickstart.rst

Quickstart
==========

This section explains how to upload data to Aito with either :doc:`CLI <cli>` or :doc:`Python SDK <sdk>`.

> Essentially, upload data to Aito can be broken down into the following steps:

Reviewer: ```suggestion
Essentially, the upload of data into Aito can be broken down into the following steps:
```

1. Infer a Table Schema :ref:`cli <cliQuickstartInferTableSchema>` | :ref:`sdk <sdkQuickstartInferTableSchema>`
2. Change the inferred schema if needed :ref:`cli <cliQuickstartChangeSchema>` | :ref:`sdk <sdkQuickstartChangeSchema>`
3. Create a table :ref:`cli <cliQuickstartCreateTable>` | :ref:`sdk <sdkQuickstartCreateTable>`
4. Convert the data :ref:`cli <cliQuickstartConvertData>` | :ref:`sdk <sdkQuickstartConvertData>`
5. Upload the data :ref:`cli <cliQuickstartUploadData>` | :ref:`sdk <sdkQuickstartUploadData>`

.. note::

  Skip steps 1, 2, and 3 if you upload data to an existing table
  Skip step 4 if you already have the data in the appropriate format for uploading or the data matches the table schema

If you don't have a data file, you can download our `example file <https://raw.githubusercontent.com/AitoDotAI/kickstart/master/reddit_sample.csv>`_ and follow the guide.

> Uploading Data with CLI

Reviewer: ```suggestion
Upload Data with the CLI
```

-----------------------

.. note::

  You can use the :ref:`Quick Add Table Operation <cliQuickAddTable>` instead of doing upload step-by-step if
  you want to upload to a new table and don't think you need to adjust the inferred schema.

The CLI supports all steps needed to upload data:

.. _cliQuickstartInferTableSchema:

:ref:`Infer a Table Schema <cliInferTableSchema>`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For examples, infer a table schema from a csv file::

  $ aito infer-table-schema csv < path/to/myCSVFile.csv > path/to/inferredSchema.json

.. _cliQuickstartChangeSchema:

Change the Schema
~~~~~~~~~~~~~~~~~

You might want to change the ColumnType_, e.g: The ``id`` column should be of type ``String`` instead of ``Int``,

> or add a Analyzer_ to a ``Text`` column. In that case, just make changes to the inferred schema JSON file.

Reviewer: ```suggestion
or add an Analyzer_ to a ``Text`` column. In that case, just make changes to the inferred schema JSON file.
```

> The example below use ``jq`` to change the ``id`` column type::

Reviewer: ```suggestion
The example below uses the ``jq`` program to change the ``id`` column type::
```

  $ jq '.columns.id.type = "String"' < path/to/schemaFile.json > path/to/updatedSchemaFile.json

.. _cliQuickstartCreateTable:

:ref:`Create a Table <cliCreateTable>`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You need a table name and a table schema to create a table::

  $ aito database create-table tableName path/to/tableSchema.json

.. _cliQuickstartConvertData:

:ref:`Convert the Data <cliConvert>`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you made change to the inferred schema or have an existing schema, use the schema when with the ``-s`` flag to make sure that the converted data matches the schema::

Reviewer: ```suggestion
If you made changes to the inferred schema or have an existing schema, use the schema when with the ``-s`` flag to make sure that the converted data matches the schema::
```

  $ aito convert csv -s path/to/updatedSchema.json path/to/myCSVFile.csv > path/to/myConvertedFile.ndjson

You can either convert the data to:

  - A list of entries in JSON format for `Batch Upload`_::

      $ aito convert csv --json path/to/myCSVFile.csv > path/to/myConvertedFile.json

  - A NDJSON file for `File Upload`_::

      $ aito convert csv < path/to/myFile.csv > path/to/myConvertedFile.ndjson

    Remember to gzip the NDJSON file::

      $ gzip path/to/myConvertedFile.ndjson

.. _cliQuickstartUploadData:

Upload the Data
~~~~~~~~~~~~~~~

  You can upload data with the CLI by using the :ref:`cliDatabase`.

  First, :ref:`cliSetUpAitoCredentials`. The easiest way is by using the environment variables::

>     $ source AITO_INSTANCE_NAME=your-instance-name

Reviewer: Again should this be export?

    $ source AITO_API_KEY=your-api-key

  You can then upload the data by either:

    - :ref:`cliBatchUpload`::

        $ aito database upload-batch tableName < tableEntries.json

    - :ref:`cliFileUpload`::

        $ aito database upload-file tableName tableEntries.ndjson.gz

> Uploading Data with SDK

Reviewer: ```suggestion
Upload Data with the SDK
```

-----------------------

The Aito Python SDK uses `Pandas DataFrame`_ for multiple operations.

> The example belows show how you can load a csv file into a DataFrame, please read the `official guide <https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html>`__ for further instructions.

Reviewer: ```suggestion
The example below show how you can load a csv file into a DataFrame, please read the `official guide <https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html>`__ for further instructions.
```

.. code:: python

  import pandas as pd

>   reddit_df = pd.read_csv('reddit_sampe.csv')

Reviewer: ```suggestion
  reddit_df = pd.read_csv('reddit_sample.csv')
```

:ref:`Infer a Table Schema <sdkInferTableSchema>`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :ref:`apiSchemaHandler` can infer table schema from a DataFrame:

  .. code:: python

    from aito.utils.schema_handler import SchemaHandler
    schema_handler = SchemaHandler()
    inferred_schema = schema_handler.infer_table_schema_from_pandas_data_frame(data_frame)

.. _sdkQuickstartChangeSchema:

Change the Schema
~~~~~~~~~~~~~~~~~

You might want to change the ColumnType_, e.g: The ``id`` column should be of type ``String`` instead of ``Int``,
or add a Analyzer_ to a ``Text`` column.

The return inferred schema from :ref:`apiSchemaHandler` is a `Python Dictionary Object`_ and hence, can be updated by updating the value:

  .. code :: python

    inferred_schema['columns']['id']['type'] = 'String'

.. _sdkQuickstartCreateTable:

:ref:`Create a Table <sdkCreateTable>`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The :ref:`apiAitoClient` can create a table using a table name and a table schema:

  .. code:: python

    from aito.utils.aito_client import AitoClient
    aito_client = AitoClient(instance_name="your_aito_instance_name", api_key="your_rw_api_key")

>     aito_client.put_table_schema(table_name='your-table-name', table_schema=table_schema_content)

Reviewer: It would be good to have an example table schema here
